### PR TITLE
Update digest field in all artifacthub-pkg.yml

### PR DESCRIPTION
**Pull request by guohaoyu110**
## Related Issue(s)
Close: https://github.com/kyverno/policies/issues/581

## Description
update digest field in all artifacthub-pkg.yml according to the contribution guide.

## Checklist

- [x] I have read the [policy contribution guidelines](https://github.com/kyverno/policies/blob/main/README.md#contribution).
- [x] I have added test manifests and resources covering both positive and negative tests that prove this policy works as intended.
- [x] I have added the artifacthub-pkg.yml file and have verified it is complete and correct.

### CODE REVIEW

**Comment by guohaoyu110** [comment]
@chipzoller  could you please look at this pr? i. don't know what's wrong with it.

**Comment by guohaoyu110** [comment]
```
#!/bin/bash

# Loop through every directory containing a .yaml file
for dir in $(find . -name "*.yaml" -exec dirname {} \; | sort -u); do
    # Get the directory name without the path
    dirname=$(basename $dir)

    # Check if a file with the same name as the directory exists in that directory
    if [ -f "$dir/$dirname.yaml" ]; then
        # Calculate the SHA256 hash of the file
        sha256=$(shasum -a 256 "$dir/$dirname.yaml" | cut -d' ' -f1)

        # Add the hash to the artifacthub-pkg.yml file in the same directory in the format "digest: <hash>"
        echo "digest: $sha256" >> "$dir/artifacthub-pkg.yml"
    fi
done
```

**Comment by guohaoyu110** [comment]
the modified workflow as below:
```
#!/bin/bash

set -euo pipefail

# Loop through each policy directory in the repository
for policy_dir in $(find "$GITHUB_WORKSPACE" -type d -name '*-*'); do
  # Validate that the folder name contains only dashes as the non-alphanumeric characters
  if [[ ! $(basename "$policy_dir") =~ ^[a-zA-Z0-9]+(-[a-zA-Z0-9]+)*$ ]]; then
    echo "Policy folder name in directory $policy_dir contains invalid characters"
    exit 1
  fi

  # Validate that policy is named the same as the folder
  policy_file=$(find "$policy_dir" -maxdepth 1 -type f -name '*.yaml')
  policy_name=$(basename "$policy_file" .yaml)

  if [[ $(basename "$policy_dir") != "$policy_name" ]]; then
    echo "Policy in directory $policy_dir is not named the same as the folder"
    exit 1
  fi

  # Validate that artifacthub-pkg.yml file is found in the same folder as the policy
  if [[ ! -f "$policy_dir/artifacthub-pkg.yml" ]]; then
    echo "artifacthub-pkg.yml file is not found in the same folder as the policy in directory $policy_dir"
    exit 1
  fi
done



```

**Comment by guohaoyu110** [comment]
```

#!/bin/bash

set -euo pipefail

# Loop through each policy directory in the repository
for policy_dir in $(find "$GITHUB_WORKSPACE" -type d -name '*-*'); do
  # Validate that the folder name contains only dashes as the non-alphanumeric characters
  if [[ ! $(basename "$policy_dir") =~ ^[a-zA-Z0-9]+(-[a-zA-Z0-9]+)*$ ]]; then
    echo "Policy folder name in directory $policy_dir contains invalid characters"
    exit 1
  fi

  # Validate that each policy file in the directory has the same parent folder name as part of its name
  policy_files=$(find "$policy_dir" -maxdepth 1 -type f -name '*.yaml')
  for policy_file in $policy_files; do
    policy_name=$(basename "$policy_file" .yaml)
    if [[ ! "$policy_name" =~ ^$(basename "$policy_dir") ]]; then
      echo "Policy $policy_file in directory $policy_dir does not have the same parent folder name"
      exit 1
    fi
  done

  # Validate that artifacthub-pkg.yml file is found in the same folder as the policy
  if [[ ! -f "$policy_dir/artifacthub-pkg.yml" ]]; then
    echo "artifacthub-pkg.yml file is not found in the same folder as the policy in directory $policy_dir"
    exit 1
  fi
done

```

**Comment by guohaoyu110** [comment]
> @chipzoller the digest field already updated, the script is already rewritten as above
```
#!/bin/bash

# Loop through every directory containing a .yaml file
for dir in $(find . -name "*.yaml" -exec dirname {} \; | sort -u); do
    # Get the directory name without the path
    dirname=$(basename $dir)

    # Check if a file with the same name as the directory exists in that directory
    if [ -f "$dir/$dirname.yaml" ]; then
        # Calculate the SHA256 hash of the file
        sha256=$(shasum -a 256 "$dir/$dirname.yaml" | cut -d' ' -f1)

        # Add the hash to the artifacthub-pkg.yml file in the same directory in the format "digest: <hash>"
        echo "digest: $sha256" >> "$dir/artifacthub-pkg.yml"
    fi
done
```

**Comment by guohaoyu110** [comment]
@chipzoller  in this pr, i have not changed the ci.yml, and i have changed all the digest field, please see the script above. I still have problems with the ci.yml, will add the updates in the next pr.

File: .github/workflows/ci.yml
Comment (chipzoller) [inline .github/workflows/ci.yml:1]: Please remove the empty new line from this file so the PR only focuses on the changes to the metadata files.

**Review by chipzoller — CHANGES_REQUESTED** [review]

File: .github/workflows/ci.yml
Comment (guohaoyu110) [reply .github/workflows/ci.yml:1]: solved. please review @chipzoller

**Comment by chipzoller** [comment]
Why was this closed? We still need the changes.

**Comment by guohaoyu110** [comment]
@chipzoller  i tested my digest value and for example `shasum -a 256 require-probes.yaml`, the digest value it produced is different from yours. Is it something wrong with the script or the shasum is different for every single computer?

**Comment by chipzoller** [comment]
No, SHA256 sum should be consistent regardless of tooling, so you need to look at the script again.

**Comment by guohaoyu110** [comment]
@chipzoller i don't know why the result is this, but when i do the shasum, the digest value is different from yours. 
Screenshot 2023-04-27 at 3 16 46 AM

**Comment by guohaoyu110** [comment]
> I looked into this and apparently the line endings are coded differently for Linux and Windows which would result in the hash being different. As long as the script is looking at the correct file, that's what matters here. And, really, Artifact Hub doesn't attempt to match the digest with the file to determine if changes occurred, it just looks at the digest to see if it's different.

ok, understand.

**Pull request closed without merging**